### PR TITLE
default binding modes: add pat_binding_modes

### PR DESCRIPTION
**Pull request by tbg**
This PR kicks off the implementation of the [default binding modes RFC][1] by
introducing the `pat_binding_modes` typeck table mentioned in the [mentoring
instructions][2].

It is a WIP because I wasn't able to avoid all uses of the binding modes as
not all call sites are close enough to the typeck tables. I added marker
comments to any line matching `BindByRef|BindByValue` so that reviewers
are aware of all of them.

I will look into changing the HIR (as suggested in [2]) to not carry a
`BindingMode` unless one was explicitly specified, but this PR is good for
a first round of comments.

The actual changes are quite small and CI will fail due to overlong lines
caused by the marker comments.

See #42640.

cc @nikomatsakis

[1]: https://github.com/rust-lang/rfcs/pull/2005
[2]: https://github.com/rust-lang/rust/issues/42640#issuecomment-313535089

### CODE REVIEW

File: src/librustc/hir/lowering.rs
Comment (nikomatsakis) [inline src/librustc/hir/lowering.rs]: Yes, this is fine, it occurs before type-check even starts.

File: src/librustc/middle/mem_categorization.rs
Comment (nikomatsakis) [inline src/librustc/middle/mem_categorization.rs]: just add an argument to the fn; it's local to this file, and the callers can use `self.tables`

File: src/librustc/middle/region.rs
Comment (nikomatsakis) [inline src/librustc/middle/region.rs]: Hmm, I think this is OK, but it's a subtle one. This has to do with when we will consider temporaries on the right-hand-side (that is, the initializer) to live for a longer time. So in particular if you write something like:

```rust
let (ref x, ref y) = (Foo { .. }, Bar { .. });
```

the temporary value created for the initializer (the tuple) would be extended to live until the end of the enclosing block (as opposed to being dropped after the `let` is complete).

To create an *implicit* ref, however, you must have a borrowed value on the RHS already. So that means you would be doing something like:

```
let Foo { x, .. } = &Foo { x: ..., ... };
```

in place of

```
let Foo { ref x, .. } = Foo { ... }
```

In the former case (the implicit ref version), the temporary is created by the `&` expression, and indeed its lifetime would be extended to the end of the block, but due to a different rule, not `is_binding_pat`.

It's probably worth copying this comment into the source as a note for why it's ok to just look for explicit `ref` keywords here.

File: src/librustc_borrowck/borrowck/mod.rs
Comment (nikomatsakis) [inline src/librustc_borrowck/borrowck/mod.rs]: I think we should use a distinct type for the values stored in the table -- i.e., `ty::BindingMode::BindByRef` or whatever. Were you considering doing that as part of this PR? It would cover a lot of these "vetted" cases -- i.e., a grep for `hir::BindingMode` would yield far fewer results. (Ideally, we would give them distinct names, too...)

Maybe `hir::BindingAnnotation` and `ty::BindingMode` or something like that? Doesn't seem great; I'm shooting for names that convey that what is in HIR is not the final mode, but just covers what the user actually wrote. I envision:

```rust
pub enum BindingAnnotation {
  // No binding annotation given: this means that the final binding mode
  // will depend on whether we have skipped through a `&` refrence
  // when matching. For example, the `x` in `Some(x)` will have binding
  // mode `None`; if you do `let Some(x) = &Some(22)`, it will ultimately be
  // inferred to be by-reference.
  None,

  // Annotated with `mut x` -- could be either ref or not, similar to `None`.
  Mutable,

  // Annotated as `ref`, like `ref x`
  Ref,

  // Annotated as `ref mut x`.
  RefMut,
}
```

File: src/librustc_lint/unused.rs
Comment (nikomatsakis) [inline src/librustc_lint/unused.rs]: Huh. That is indeed confusing; it would be helpful if you did something like this:

```
p.each_binding(|bm, id, span, path1| { // <-- I added `span` in place of `_`
  let bm = match cx.tables.pat_bindings_modes.get(&p.id) {
    Some(&bm) => bm,
    None => span_bug!(span, "missing binding mode")
  };
  ...
```

**Review by nikomatsakis — COMMENTED** [review]
Mostly looks great! I left suggestions at all the problematic places (I think!).

**Comment by alexcrichton** [comment]
ping @tschottdorf, just wanted to make sure this is still on your radar!

**Comment by tschottdorf** [comment]
It is, no worries :) my open source day is Friday, though.

**Review by tbg — COMMENTED** [review]
@nikomatsakis ready for another look. I kept the changes in a second commit for easier reviewing on your end, but the intention is that it would be squashed before merge.

File: src/librustc/hir/mod.rs
Comment (nikomatsakis) [inline src/librustc/hir/mod.rs]: Nit: Nice comment. Would be better to use `///`

File: src/librustc/middle/mem_categorization.rs
Comment (nikomatsakis) [inline src/librustc/middle/mem_categorization.rs]: Nit: remove this `FIXME`

File: src/librustc_lint/unused.rs
Comment (nikomatsakis) [inline src/librustc_lint/unused.rs]: did we ever get to the bottom of this? I guess so...?

File: src/librustc_typeck/check/_match.rs
Comment (nikomatsakis) [inline src/librustc_typeck/check/_match.rs]: Yeah, this is annoying. The comments in #23116 are somewhat helpful in bringing this back into cache. I have a feeling that scraping the HIR is actually just fine, but I have to think about it and convince myself (and if not, decide what to do...).

**Review by nikomatsakis — APPROVED** [review]
Looks good to me! Left a nit but I don't feel inclined to block this PR on it.

File: src/librustc_lint/unused.rs
Comment (tschottdorf) [inline src/librustc_lint/unused.rs]: Yeah, I had used id instead of p.id.

**Comment by kennytm** [comment]
@bors retry

`x86_64-pc-windows-msvc` failed without any log. [Corresponding Travis build](https://travis-ci.org/rust-lang/rust/builds/259011895) has two successful jobs, so should be spurious.

File: src/librustc_typeck/check/_match.rs
Comment (arielb1) [inline src/librustc_typeck/check/_match.rs]: No they're not. The problem is that if we have something like
```Rust
    let &mut Foo(x) = f()[0];
```

Then if the pattern matches by reference, we want to match `f()[0]` as a lexpr, so we can't allow it to be coerced. But if the pattern matches by value, `f()[0]` is still syntactically a lexpr, but we *do* want to allow coercions.

Comment (arielb1) [inline src/librustc_typeck/check/_match.rs]: I *think* we can be okay with allowing coercions to happen if there are no explicit `ref mut` patterns - all *implicit* `ref mut` patterns must occur behind a reference, so they will have the "correct" variance and lifetime.

Comment (arielb1) [inline src/librustc_typeck/check/_match.rs]: This means that the following pattern would be legal:
```Rust
struct Foo(Bar);
struct Bar(u32);
impl Deref for Foo {
    type Target = Bar;
    fn deref(&self) -> &Bar { &self.0 }
}
impl DerefMut for Foo {
    fn deref_mut(&mut self) -> &mut Bar { &mut self.0 }
}
fn foo(x: &mut Foo) {
    {
        let Bar(z) = x;
        *z = 42;
    }
    assert_eq!(foo.0.0, 42);
}
```

**Comment by tschottdorf** [comment]
https://ci.appveyor.com/project/rust-lang/rust/build/1.0.4120/job/nef70pywepcisqra

```
   Compiling rls-span v0.4.0
   Compiling languageserver-types v0.12.0
error[E0432]: unresolved import `syntax::ast::BindingMode`
  --> C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\rustfmt-nightly-0.2.1\src\patterns.rs:11:25
   |
11 | use syntax::ast::{self, BindingMode, FieldPat, Pat, PatKind, RangeEnd};
   |                         ^^^^^^^^^^^ no `BindingMode` in `ast`
error[E0433]: failed to resolve. Could not find `BindingMode` in `ast`
    --> C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\rustfmt-nightly-0.2.1\src\items.rs:1725:37
     |
1725 |     if let ast::PatKind::Ident(ast::BindingMode::ByValue(mutability), _, _) = arg.pat.node {
     |                                     ^^^^^^^^^^^ Could not find `BindingMode` in `ast`
   Compiling syntex_errors v0.52.0
error: aborting due to 2 previous errors
error: Could not compile `rustfmt-nightly`.
warning: build failed, waiting for other jobs to finish...
error: build failed

```

**Comment by tschottdorf** [comment]
I seem to have accidentally renamed `ast::BindingMode` to `ast::BindingAnnotation` (one could argue that the AST should change similar to how `hir::BindingMode` has changed, but definitely not in this PR). Fixing.

**Comment by tschottdorf** [comment]
Cleaned up.

**Comment by arielb1** [comment]
@tschottdorf 

It would be nice if you would add my comment about binding modes as a code comment.

**Comment by tschottdorf** [comment]
@arielb1 happy to do so, could you provide the comment? I'm not sure what exactly you'd want to see there at this point.

File: src/librustc_typeck/check/_match.rs
Comment (nikomatsakis) [inline src/librustc_typeck/check/_match.rs]: > I think we can be okay with allowing coercions to happen if there are no explicit ref mut patterns - all implicit ref mut patterns must occur behind a reference, so they will have the "correct" variance and lifetime.

Yes, this was my thinking. I think that the current rules are "stricter than necessary" in this respect (as the comments indeed indicate).

> This means that the following pattern would be legal:

I am not sure what your code is showing. In particular, was `assert_eq!(foo.0.0, 42);` meant to be `x.0.0`? Also, did you envision a problem with this code being accepted? (I would expect it to work, naively.)

**Comment by kennytm** [comment]
@bors force retry (let's see if this works)

bors is stuck waiting for 43009 to build which has been merged 2 days ago...?

**Pull request merged**
